W-11419023-Persistent-queues-v4.4-DM (#2328)
mulesoft/docs-mule-runtime: modules/ROOT/pages/reliability-patterns.adoc

If you have a web application that uses a nontransactional connector, such as HTTP, follow a reliability pattern to ensure reliable messaging for your application.

== Comparing Endpoints in Reliability Patterns

You can couple the reliable acquisition flow with any type of transactional endpoint. You don't have to use a VM endpoint in a reliability pattern. The following diagram illustrates a reliability pattern where the reliable acquisition flow delivers a message to a JMS endpoint:

image::reliabilitypatternwithjms.png[ReliabilityPatternwithJMS]

The following table lists the advantages and disadvantages of using VM, JMS, and JDBC endpoints in a reliable acquisition flow.

[%header,cols="10a,60a,30a"]
|===
|Endpoint |Advantage |Disadvantage
|File-based VM (standalone Mule)
|
* Provided ready to use with Mule. There is no need for additional software.
|

> * Only applies to non-clustered environments.

Reviewer: ```suggestion
* Applies only to non-clustered environments.
```

|JDBC
|
* Depending on resources available, JDBC can hold a large amount of information.
|
* Requires additional software (database server).
* Can be complex to use since proper DBMS schemas are required.
|JMS (enabled for persistence)
|

> * It is the most reliable choice because messages are persisted to disk. It applies to clustered environments.

Reviewer 2: ```suggestion
* Applies to both clustered and non-clustered environments.
```

|
* Requires additional software (message queue server).
|===

[NOTE]

> In memory persistence setups for both VM and JMS are excluded from the comparison since they can lose messages; therefore, they are not considered reliable.

Reviewer: ```suggestion
The comparison doesn't include in-memory persistence setups for VM and JMS because they can lose messages and are not considered reliable.
```
